### PR TITLE
FIX: add support for desc order when ranking infs with nans #19538

### LINKED ISSUE #19538: Rank Mixes np.nan with np.inf values
#### Code Sample, a copy-pastable example if possible

```python
In []: df = pd.DataFrame([1, np.nan, np.inf, -np.inf, 25])
In []: df.rank()
Out []:
     0
0  2.0
1  NaN
2  NaN
3  1.0
4  3.0

In []: df.rank(ascending=False)
Out []:
     0
0  3.0
1  NaN
2  1.0
3  NaN
4  2.0
```
#### Problem description

``np.inf`` or ``-np.inf`` gets grouped with ``np.nan`` in the rank operation, depending on which direction the ranking occurs in. Ideally, ``np.inf`` would be entirely separate from ``np.nan``.

### PR DESCRIPTION
**Pull request by peterpanmj**
Please include the output of the validation script below between the "```" ticks:

```
################################################################################
################ Docstring (pandas._libs.algos.rank_1d_object)  ################
################################################################################

Fast NaN-friendly version of scipy.stats.rankdata

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
        Summary does not end with dot
        No extended summary found
        No returns section found
        See Also section not found
        No examples section found
(pandas_dev)

```

Checklist for other PRs (remove this part if you are doing a PR for the pandas documentation sprint):

- [x] closes #19538
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/series/test_rank.py [hunk 1]
```diff
@@ -263,8 +264,11 @@ def test_rank_tie_methods_on_infs_nans(self):
         chunk = 3
         disabled = set([('object', 'first')])
 
-        def _check(s, expected, method='average', na_option='keep'):
-            result = s.rank(method=method, na_option=na_option)
+        def _check(s, expected, method='average', na_option='keep',
```
Comment (jreback) [inline pandas/tests/series/test_rank.py]: can you parametrize this test

File: pandas/tests/series/test_rank.py [hunk 2]
```diff
@@ -283,12 +287,13 @@ def _check(s, expected, method='average', na_option='keep'):
...
-                    order = ranks[0] + ranks[2] + ranks[1]
+                    order = [ranks[0], ranks[2], ranks[1]]
                 else:
-                    order = ranks[0] + [np.nan] * chunk + ranks[1]
-                _check(iseries, order, method, na_opt)
+                    order = [ranks[0], [np.nan] * chunk, ranks[1]]
+                _check(iseries, order, method, na_opt, True)
+                _check(iseries, order[::-1], method, na_opt, False)
 
     def test_rank_methods_series(self):
```
Comment (jreback) [inline pandas/tests/series/test_rank.py]: ca you change this to use the ``@td.skip_if_no_scipy`` decorator instead

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a whatsnew note

**Comment by jreback** [comment]
can you add the simple example from the issue as a test as well.

**Comment by peterpanmj** [comment]
> The command "ci/lint.sh" exited with 1.

The Travis CI build failed after some modification on "pandas\tests\series\test_rank.py". I have no idea what is the cause.

**Comment by jreback** [comment]
somehow lots of commits go there. merge in master and push again.

File: pandas/tests/series/test_rank.py [hunk 3]
```diff
@@ -257,38 +259,50 @@ def _check(s, expected, method='average'):
...
         exp_ranks = {
             'average': ([2, 2, 2], [5, 5, 5], [8, 8, 8]),
             'min': ([1, 1, 1], [4, 4, 4], [7, 7, 7]),
             'max': ([3, 3, 3], [6, 6, 6], [9, 9, 9]),
             'first': ([1, 2, 3], [4, 5, 6], [7, 8, 9]),
             'dense': ([1, 1, 1], [2, 2, 2], [3, 3, 3])
         }
-        na_options = ('top', 'bottom', 'keep')
+
+        def _check(s, method, na_option, ascending):
```
Comment (jreback) [inline pandas/tests/series/test_rank.py]: can you inline these below

File: pandas/tests/series/test_rank.py [hunk 4]
```diff
@@ -257,38 +259,50 @@ def _check(s, expected, method='average'):
...
-                    order = ranks[0] + ranks[2] + ranks[1]
-                else:
-                    order = ranks[0] + [np.nan] * chunk + ranks[1]
-                _check(iseries, order, method, na_opt)
+            if (dtype, method) in disabled:
+                continue
+            _check(iseries, method, na_option, ascending)
+
+    def test_rank_desc_mix_nans_infs(self):
+        iseries = Series([1, np.nan, np.inf, -np.inf, 25])
```
Comment (jreback) [inline pandas/tests/series/test_rank.py]: can you add the issue number

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by peterpanmj** [comment]
lint.sh gives error. Don't know why.

File: pandas/tests/series/test_rank.py [hunk 5]
```diff
@@ -257,38 +259,52 @@ def _check(s, expected, method='average'):
...
         dtypes = [('object', None, Infinity(), NegInfinity()),
                   ('float64', np.nan, np.inf, -np.inf)]
         chunk = 3
         disabled = set([('object', 'first')])
 
-        def _check(s, expected, method='average', na_option='keep'):
-            result = s.rank(method=method, na_option=na_option)
+        def _check(s, method, na_option, ascending):
+            exp_ranks = {
+                    'average': ([2, 2, 2], [5, 5, 5], [8, 8, 8]),
```
Comment (TomAugspurger) [inline pandas/tests/series/test_rank.py]: These should start four spaces to the right of the `e` in `exp_ranks`.

Comment (TomAugspurger) [reply pandas/tests/series/test_rank.py]: Likewise with each one below it.

File: pandas/tests/series/test_rank.py [hunk 6]
```diff
@@ -257,38 +259,52 @@ def _check(s, expected, method='average'):
...
-                    order = ranks[0] + ranks[2] + ranks[1]
-                else:
-                    order = ranks[0] + [np.nan] * chunk + ranks[1]
-                _check(iseries, order, method, na_opt)
+            if (dtype, method) in disabled:
+                continue
+            _check(iseries, method, na_option, ascending)
+
+    def test_rank_desc_mix_nans_infs(self):
+        #GH 19538
```
Comment (TomAugspurger) [inline pandas/tests/series/test_rank.py]: Space after `#`

File: pandas/tests/series/test_rank.py [hunk 7]
```diff
@@ -257,38 +259,52 @@ def _check(s, expected, method='average'):
...
+        #check descending ranking when mix nans and infs
```
Comment (TomAugspurger) [inline pandas/tests/series/test_rank.py]: Space after `#`

**Review by TomAugspurger — COMMENTED** [review]
Added comments on the linking errors in https://travis-ci.org/pandas-dev/pandas/jobs/353646265#L3004

Run `flake8 pandas/tests/series/test_rank.py` locally before pushing to validate your fixes.

**Comment by jreback** [comment]
@WillAyd if you'd have a look

**Review by jreback — APPROVED** [review]

**Comment by WillAyd** [comment]
@peterpanmj nice work - does this work for GroupBy ranking the same way?

**Comment by peterpanmj** [comment]
@WillAyd  As long as rank_1d_{dtype} is called, it is fine, but group by rank is still broken.
```
In [51]: df = pd.DataFrame([1, np.nan, np.inf, -np.inf, 25])

In [52]: df['key'] = 'foo'

In [53]: df.groupby("key").rank()      # not working properly for infinity
Out[53]:
     0
0  2.0
1  NaN
2  NaN
3  1.0
4  3.0

In [54]: df.rank()     # this is ok
Out[54]:
     0  key
0  2.0  3.0
1  NaN  3.0
2  4.0  3.0
3  1.0  3.0
4  3.0  3.0

In [55]: df.groupby("key").apply(lambda x:x.rank())   # this is also ok
Out[55]:
     0  key
0  2.0  3.0
1  NaN  3.0
2  4.0  3.0
3  1.0  3.0
4  3.0  3.0

```

**Comment by WillAyd** [comment]
Do you want to try and tackle the groupby implementation as part of this? I've linked the relevant function below - it isn't that drastically different from the algos rank implementation

https://github.com/pandas-dev/pandas/blob/766a4800b4d9b7e847577b74b24128fcda8eeb55/pandas/_libs/groupby_helper.pxi.in#L415

**Comment by peterpanmj** [comment]
I will try but it should take a lot of efforts. 
I don't under stand why _out_ has ndim=2 ? What is the dtype of _labels_ ?

**Comment by WillAyd** [comment]
Labels are int64 - each unique grouping will have it's own label. out has `ndim=2` to match the other Cython call signatures, but you'll see in this that only the first dimension ever gets written to (I think this will be cleaned up in the future).

**Comment by jreback** [comment]
can do the groupby fixes in another PR :> @peterpanmj or @WillAyd can you open an issue?

**Comment by jreback** [comment]
thanks @peterpanmj

**Pull request merged**